src/cluster_migrate.c

```diff
@@ -0,0 +1,2003 @@
+    listNode *ln;
+    listIter li;
+    if (sdslen(name) != CLUSTER_NAMELEN) return NULL;
+    listRewind(server.cluster->slot_migration_jobs, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        slotMigrationJob *job = ln->value;
+        if (!memcmp(name, job->name, CLUSTER_NAMELEN)) return job;
+    }
+    return NULL;
+}
+
+/* Cancels one or all ongoing imports. */
+void clusterCommandCancelMigration(client *c) {
+    listNode *ln;
+    listIter li;
+
+    if (!strcasecmp(c->argv[2]->ptr, "all") && c->argc == 3) {
+        if (!clusterIsAnySlotExporting()) {
+            addReplyError(c, "No migrations ongoing");
+            return;
+        }
+
+        listRewind(server.cluster->slot_migration_jobs, &li);
+        while ((ln = listNext(&li)) != NULL) {
+            slotMigrationJob *job = ln->value;
+            if (!isSlotMigrationJobInProgress(job) || job->type == SLOT_MIGRATION_IMPORT) {
+                continue;
+            }
+            finishSlotMigrationJob(job, SLOT_MIGRATION_JOB_CANCELLED, NULL);
+        }
+        serverLog(LL_NOTICE, "Canceled all in progress slot migrations due to operator's request.");
+        addReply(c, shared.ok);
+        return;
+    } else if (!strcasecmp(c->argv[2]->ptr, "name") && c->argc > 3) {
+        slotMigrationJob *job = clusterLookupMigrationJob(c->argv[3]->ptr);
+        if (!job || !isSlotMigrationJobInProgress(job)) {
+            addReplyErrorFormat(c, "No outgoing migration with name found.");
+            return;
+        }
+        if (job->type == SLOT_MIGRATION_IMPORT) {
+            addReplyErrorFormat(c, "Migrations must be cancelled on the node that currently owns the slots.");
+            return;
+        }
+        serverLog(LL_NOTICE,
+                  "Cancelling slot migration %s due to operator's request.",
+                  job->description);
+        finishSlotMigrationJob(job, SLOT_MIGRATION_JOB_CANCELLED, NULL);
+        addReply(c, shared.ok);
+        return;
+    } else {
+        addReplyErrorObject(c, shared.syntaxerr);
+        return;
+    }
+}
+
+/* Handler for connect callbacks during slot migration job connection. */
+void slotExportConnectHandler(connection *conn) {
+    slotMigrationJob *job = (slotMigrationJob *)connGetPrivateData(conn);
+    proceedWithSlotMigration(job);
+}
+
+/* Connect the given job to the target node. The created connection will have the job as private
+ * data. */
+void connectSlotExportJob(slotMigrationJob *job) {
+    sds status_msg;
+    clusterNode *n = clusterLookupNode(job->nodename, CLUSTER_NAMELEN);
+    int port = getNodeDefaultReplicationPort(n);
+    serverLog(LL_NOTICE, "Connecting slot migration %s (ip: %s, port %d)",
+              job->description,
+              n->ip,
+              port);
+
+    job->conn = connCreate(connTypeOfReplication());
+    if (connConnect(job->conn, n->ip, port, server.bind_source_addr,
+                    /*multipath=*/0, slotExportConnectHandler) == C_ERR) {
```

Comment:
let's remove `/*multipath=*/`, we don't use it a lot i think (i also don't like it very much TBH)